test/gtest/ucp/test_ucp_tag.cc

```diff
@@ -464,16 +466,89 @@ UCP_INSTANTIATE_TEST_CASE(test_ucp_tag_limits)
 
 class test_ucp_tag_nbx : public test_ucp_tag {
 public:
+    class completion_value {
+    public:
+        completion_value() : m_address(nullptr), m_value(0)
+        {
+        }
+
+        completion_value(uint32_t *address, uint32_t value) :
+            m_address(address), m_value(value)
+        {
+        }
+
+        bool empty() const
+        {
+            return (m_address == nullptr) && (m_value == 0);
+        }
+
+        uint32_t *m_address;
+        uint32_t  m_value;
+    };
+
+    test_ucp_tag_nbx()
+    {
+        if (proto_enbaled()) {
+            modify_config("PROTO_ENABLE", "y");
+        }
+    }
+
     void init() {
-        /* forbid zcopy access because it will always fail due to read-only
-         * memory pages (will fail to register memory) */
-        modify_config("ZCOPY_THRESH", "inf");
+        stats_activate();
         test_ucp_tag::init();
     }
 
+    virtual void cleanup()
+    {
+        test_ucp_tag::cleanup();
+        stats_restore();
+    }
+
+    static void
+    get_test_variants_prereg(std::vector<ucp_test_variant> &variants)
+    {
+        add_variant_with_value(variants, UCP_FEATURE_TAG, 0, "");
+        add_variant_with_value(variants, UCP_FEATURE_TAG, TEST_FLAG_PREREG,
+                               "prereg");
+    }
+
+    static void get_test_variants(std::vector<ucp_test_variant> &variants)
+    {
+        add_variant_values(variants, get_test_variants_prereg, 0);
+        add_variant_values(variants, get_test_variants_prereg, TEST_FLAG_PROTO,
+                           "proto");
+        add_variant_values(variants, get_test_variants_prereg, TEST_FLAG_IOV,
+                           "iov");
+        add_variant_values(variants, get_test_variants_prereg,
+                           TEST_FLAG_PROTO | TEST_FLAG_IOV, "proto_iov");
+    }
+
 protected:
     static const size_t MSG_SIZE;
     static const ucp_request_param_t null_param;
+    static const completion_value null_completion;
+    static const size_t iov_count;
+
+    enum {
+        TEST_FLAG_PREREG = UCS_BIT(0),
+        TEST_FLAG_PROTO  = UCS_BIT(1),
+        TEST_FLAG_IOV    = UCS_BIT(2)
+    };
+
+    bool prereg() const
+    {
+        return get_variant_value(0) & TEST_FLAG_PREREG;
+    }
+
+    bool is_iov() const
+    {
+        return get_variant_value(1) & TEST_FLAG_IOV;
```

Comment:
why using variant 1 and not 0?
it's weird that TEST_FLAG_PREREG, TEST_FLAG_PROTO, TEST_FLAG_IOV are exclusive flags but using different variant index